### ISSUE TITLE
tensor.rand() and uniform_() returns numbers from [0,1] (right-hand inclusive)

### ISSUE THREAD

**Issue by janchorowski** [pytorch/pytorch#2740]
It is customary that the uniform random generator returns numbers from a range that includes the lower bound, but excludes the upper bound. This is the behavior of e.g. numpy (https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.random.rand.html#numpy-random-rand). This is also stated in the torch.rand documentation (http://pytorch.org/docs/master/torch.html#torch.rand). However, using either the `torch.rand()` generator or the in-place `uniform_()` generator ones are sampled quite often:
```
>>> torch.__version__
'0.2.0_4'
>>> torch.rand(100000000).max()
1.0
>>> torch.FloatTensor(100000000).uniform_().max()
1.0
>>> torch.cuda.FloatTensor(100000000).uniform_().max()
1.0
```

This is useful e..g for generation of numbers that exclude the upper bound is useful e.g. to generate uniformly distributed random integers. With the current generator one has to be careful to handle the incorrect upper bound being sampled.

**Comment by ngimel** [comment]
Don't know about cpu-side generation, but on the GPU side curand_uniform is generating numbers where 0 is excluded and 1 is included: http://docs.nvidia.com/cuda/curand/device-api-overview.html#distributions

**Comment by janchorowski** [comment]
Thanks!

I haven't looked into curand docs and relied on the torch documentation (still learning it). I just checked that the CPU `torch.rand` and `torch.FloatTensor.uniform_` do return 0 and 1 occasionally, while the CUDA `torch.cuda.FloatTensor.uniform_` returns 1, but not 0.

I guess either convention is fine as long as it is documented and consistent. For now, the only exact bound specification is given for `torch.rand()` and it excludes 1.

**Comment by colesbury** [comment]
I think it would be good to change `uniform_` and `torch.rand` to return in the `[0, 1)` range. As @janchorowski pointed out, this is the convention for most random number APIs including Python, NumPy, C++11, Java, Go, and Julia.

R and MATLAB use the `(0, 1)` range.

**Comment by SsnL** [comment]
I thought the bug is in backend functions, but it doesn't seem to be. After some quick investigation, I found something interesting: when tested in the above fashion
1. CPU `THTensor_(randn)` never returns 1 on C side.
2. CUDA double version of uniform never returns either 0 or 1.
3. CUDA float version of uniform isn't fixed by simply changing `x` to `(1-x)` in https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorRandom.cu#L133-L134 .

So where did things go wrong? Is it when transitioning values back from C/CUDA to Python?

**Comment by ngimel** [comment]
Here's a small test, reproducing device- and host-side generation, as implemented in TH/THC:
```
#include <curand_globals.h>
#include <limits.h>
#include <iostream>
#include <iomanip>

int main(){
   uint x = UINT_MAX;
   double d = x * CURAND_2POW32_INV_DOUBLE + (CURAND_2POW32_INV_DOUBLE/2.0f);
   std::cout << std::setprecision(14) << "double max " << d << "\n";
   float f = x * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
   std::cout << "float max " << f << "\n";
//TH backend
   double dc = UINT_MAX * (1.0/4294967296.0);   
   std::cout << std::setprecision(14) << "double max cpu " << dc << "\n"; 
   float fc = (float)(UINT_MAX * (1.0/4294967296.0));
   std::cout << std::setprecision(14) << "float max cpu " << fc << "\n"; 
  
}
```
Only cpu-side double generation is doing what is promised: [0, 1). cpu-side for floats of course rounds max to 1, thus generating [0 1]
gpu-side for floats generates (0,1], as promised by curand manual, but that's not what torch manual specifies. 
gpu-side for doubles has a bug, thus generates (0,1), contrary to what's promised in the manual. 
Replacing x by (1-x) is expected to fail because of the different density of fp numbers near 0 and near 1 (thus, when 1-x is calculated for a small non-zero x, result will be rounded to 1, so (0,1] range will essentially be mapped to [0 1] range).

**Comment by SsnL** [comment]
Thanks for the explanation @ngimel . Do C side doubles always get rounded to 14 precision in pytorch? If so, should we allow float rand/uniform to output 1?

I also never seem to get 0 from `torch.rand(100000000).min()`. Do you know the reason?

For the CUDA side, would `x != 1 ? x : 0` be a reasonable thing to do?

**Comment by ngimel** [comment]
For the cpu side, there are 2^32 (approx 4.3e9) numbers that can be produced with equal probability by a call to rand() (well, TH calls its own generator, but still it is supposed to produce numbers between 0 and 2^32-1), only a single one of them (0) maps to 0. You are producing 1e8 numbers, so it is possible that there's not a single 0 in your sample. cuda side never produces 0-s (it is not supposed to).
Using your fix for cuda will introduce subtle distortions, but whatever, the definition of what it means to generate uniformly distributed floating point number is not precisely set anyway (for a fun read, see https://lemire.me/blog/2017/02/28/how-many-floating-point-numbers-are-in-the-interval-01/). There still won't be any 0-s for cuda-generated doubles because of curand bug.

**Comment by Smerity** [comment]
FYI I ran into this bug when I was generating random numbers in the C style using CPU. PyTorch docs promise [0, 1) so I multiplied `rand()` by N and expected it never to actually produce N. Rarely my program would crash - very rarely - i.e. 1 in a million as seen below ;)

```
>>> sum(((torch.rand(1000000) * 10).int() == 10).sum() for _ in range(10))
1
```

I understand the issue is horrendously complex thanks to @ngimel's fascinating commentary. For now, multiplying by some arbitrary scalar (`0.999`) prevents my issue above though does slightly impact random - but I didn't need it to be perfect anyway.

**Comment by soumith** [comment]
closed via #3541 and #3547

**Comment by mirceamironenco** [comment]
This seems to be occurring again. Sampling using 0.4.1. and 0.3.0:

```
>>> import torch
>>> torch.__version__
'0.4.1'
>>> torch.rand(100000000).max()
tensor(1.0000)
>>> torch.FloatTensor(100000000).uniform_().max()
tensor(1.0000)
>>> torch.cuda.FloatTensor(100000000).uniform_().max()
tensor(1.0000, device='cuda:0')

>>> torch.__version__
'0.3.0'
>>> torch.rand(100000000).max()
0.9999999403953552
>>> torch.FloatTensor(100000000).uniform_().max()
0.9999999403953552
>>> torch.cuda.FloatTensor(100000000).uniform_().max()
0.9999999403953552
